### PR TITLE
Well known

### PR DESCRIPTION
**Pull request by dvoet**
* use .well-known/openid-configuration to figure out the right URLs to use for oauth
* the revoke api changed on the uchicago side, fixed it
* removed config.ini and app.yaml so we don't accidentally commit rendered config files
* made redirect_uri required

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: oauth_adapter.py [hunk 1]
```diff
@@ -35,4 +39,62 @@ def refresh_access_token(self, refresh_token_str):
         """
         token_dict = {'refresh_token': refresh_token_str}
         oauth = OAuth2Session(self.client_id, token=token_dict)
-        return oauth.refresh_token(self.token_url, auth=self.basic_auth)
+        return oauth.refresh_token(self._get_token_info_url(), auth=self.basic_auth)
+
+    def revoke_refresh_token(self, refresh_token):
```
Comment (dvoet) [inline oauth_adapter.py]: moved revoke_refresh_token because it is not specific to fence but part of oauth

File: config.ini.ctmpl [hunk 2]
```diff
@@ -4,9 +4,8 @@
 [fence]
 CLIENT_ID={{ $fenceSecrets.Data.client_id }}
 CLIENT_SECRET={{ $fenceSecrets.Data.client_secret }}
-AUTHZ_URI={{if eq $environment "prod"}}https://dcp.bionimbus.org/user/oauth2/authorize{{else}}https://dcp.bionimbus.org/user/oauth2/authorize{{end}}
-REDIRECT_URI={{if eq $environment "prod"}}https://portal.firecloud.org/#fence-callback{{else if eq $environment "alpha"}}https://firecloud.dsde-alpha.broadinstitute.org/#fence-callback{{else if eq $environment "staging"}}https://firecloud.dsde-staging.broadinstitute.org/#fence-callback{{else if eq $environment "dev"}}https://firecloud.dsde-dev.broadinstitute.org/#fence-callback{{else}}http://local.broadinstitute.org/#fence-callback{{end}}
-TOKEN_URL={{if eq $environment "prod"}}https://dcp.bionimbus.org/user/oauth2/token{{else}}https://dcp.bionimbus.org/user/oauth2/token{{end}}
+OPEN_ID_CONFIG_URL={{if eq $environment "prod"}}https://dcp.bionimbus.org/user/.well-known/openid-configuration{{else}}https://dcp.bionimbus.org/user/.well-known/openid-configuration{{end}}
+USER_NAME_PATH_EXPR=/context/user/name
```
Comment (dvoet) [inline config.ini.ctmpl]: thinking forward to having this work with other oauth providers I made this configurable and tested with google.

File: oauth_adapter.py [hunk 3]
```diff
@@ -35,4 +39,47 @@ def refresh_access_token(self, refresh_token_str):
...
+        if not open_id_config:
+            open_id_config_response = urlfetch.fetch(self.open_id_config_url)
+            if open_id_config_response.status_code != 200:
+                raise endpoints.InternalServerErrorException(
+                    message='open_id_config_url [{}] returned status {}: {}'.format(self.open_id_config_url,
+                                                                                    open_id_config_response.status_code,
+                                                                                    open_id_config_response.content))
+            else:
+                open_id_config = json.loads(open_id_config_response.content)
+                memcache.add(namespace="OauthAdapter", key=self.provider_name, value=open_id_config)
```
Comment (gpolumbo-broad) [inline oauth_adapter.py]: I know it's really unlikely to change, but just in case, should we put like a once-a-day expiry on this?

Comment (dvoet) [reply oauth_adapter.py]: I was thinking we could always just go into the google console and evict if we need to... but sure

**Pull request merged**